### PR TITLE
HW9 Tests

### PR DESCRIPTION
**Pull request by DmitriyBabii**

### CODE REVIEW

File: src/test/java/com/alevel/cw3/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Test
+    public void calculate_positive() throws Exception {
+        //Assert.assertEquals(2, Main.calculate("4 - 2"));
+        assertEquals(6.0, Main.calculate("4 + 2"), 0.0);
+        assertEquals(2.0, Main.calculate("4 - 2"), 0.0);
+        assertEquals(4.0, Main.calculate("2 * 2"), 0.0);
+        assertEquals(1.0, Main.calculate("2 / 2"), 0.0);
+    }
```
Comment (Fabolo12) [inline src/test/java/com/alevel/cw3/MainTest.java]: It is better to write one test for each type of operation

**Review by Fabolo12 — CHANGES_REQUESTED** [review]
Ok, but I want to see the use of Mocks.
Refactor 
```
calculate(Scanner in) {
String expression = in.nextLine();
...
}
```
and in tests create a mock scanner

**Comment by DmitriyBabii** [comment]
Can`t mock Scanner because it final. So i mocked InputOutput class with Scanner inside.

File: src/test/java/com/alevel/cw3/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    @Test(expected = Exception.class)
+    public void numberSum_negative() throws Exception {
+        InputOutput in = Mockito.mock(InputOutput.class);
+        Mockito.when(in.getInput()).thenReturn(-1);
+        int number = in.getInput();
+
+        Main.numberSum(number);
+    }
```
Comment (Fabolo12) [inline src/test/java/com/alevel/cw3/MainTest.java:28]: This is a misuse of mock, it must be passed to the method

File: src/test/java/com/alevel/cw3/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+
+    //Tests with Scanner argument
+    @Test
+    public void calculate_plus_WithScannerArgument() throws Exception {
+        InputOutput in = Mockito.mock(InputOutput.class);
+        Mockito.when(in.getString()).thenReturn("4 + 2 + 2");
+
+        assertEquals(8.0, Main.calculate(in), 0.0);
+    }
```
Comment (Fabolo12) [inline src/test/java/com/alevel/cw3/MainTest.java:79]: This is the correct use of mock

**Review by Fabolo12 — APPROVED** [review]
10

File: src/test/java/com/alevel/cw3/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+import java.util.Scanner;
+
+import static org.junit.Assert.*;
+
+public class MainTest {
+
+    @Test
+    public void numberSum_positive() throws Exception {
+        InputOutput in = Mockito.mock(InputOutput.class);
```
Comment (Fabolo12) [inline src/test/java/com/alevel/cw3/MainTest.java:14]: You can simplify the code if you write like this
image

**Pull request merged**